File: 02_workspace/python_package_example/python_package_example/simple_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SimplePublisher(Node):
    """
    简单的发布者节点示例
    演示如何在ROS2包中创建发布者
    """
    
    def __init__(self):
        super().__init__('simple_publisher')
        
        self.publisher_ = self.create_publisher(String, 'example_topic', 10)
        
        timer_period = 1.0
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
        self.counter = 0
        
        self.get_logger().info('简单发布者节点已启动')
    
    def timer_callback(self):
        msg = String()
        msg.data = f'来自Python包的消息 #{self.counter}'
        
        self.publisher_.publish(msg)
        self.get_logger().info(f'发布: "{msg.data}"')
        
        self.counter += 1


def main(args=None):
    rclpy.init(args=args)
    node = SimplePublisher()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 02_workspace/cpp_package_example/src/simple_publisher.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

/**
 * 简单的C++发布者节点示例
 * 演示如何在ROS2 C++包中创建发布者
 */
class SimplePublisher : public rclcpp::Node
{
public:
  SimplePublisher()
  : Node("simple_publisher"), counter_(0)
  {
    publisher_ = this->create_publisher<std_msgs::msg::String>("example_topic", 10);
    
    timer_ = this->create_wall_timer(
      1000ms, std::bind(&SimplePublisher::timer_callback, this));
    
    RCLCPP_INFO(this->get_logger(), "C++简单发布者节点已启动");
  }

private:
  void timer_callback()
  {
    auto message = std_msgs::msg::String();
    message.data = "来自C++包的消息 #" + std::to_string(counter_);
    
    publisher_->publish(message);
    RCLCPP_INFO(this->get_logger(), "发布: '%s'", message.data.c_str());
    
    counter_++;
  }
  
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  size_t counter_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SimplePublisher>());
  rclcpp::shutdown();
  return 0;
}
